Add autoFill and highlightOnFocus as directive options

turt2live opened the pull request:
The readme claims that displayZeros is supported, but does not work. This adds a useful alternative to displayZeros through two options: autoFill and highlightOnFocus.

autoFill simply populates the field with `0` if there has been no number entered. This is done before formatting takes place just in case something needs to happen (such as the result being `0.00`). This also occurs before the minimum is considered, so that the minimum can be applied to autoFill.

highlightOnFocus addresses #118 and is enabled by default. This is to make entering a large number of fields easier for the user, and just generally make the field more usable.

turt2live committed:
Add autoFill and highlightOnFocus

turt2live committed:
Update README

turt2live committed:
Flip highlightOnFocus to represent expected behaviour

turt2live commented:
Looks like I blew up the tests (something that doesn't happen on my end). Will take a look.

turt2live committed:
Fix tests for highlightOnFocus

nick-woodward commented on test/ng-currency/ng-currency-settings.provider.spec.js:
```diff
@@ -18,7 +18,9 @@ describe('provider(ngCurrencySettings)', () => {
-        currencySymbol: undefined
+        currencySymbol: undefined,
+        autoFill: false,
+        highlightOnFocus: true
```
Rather then reusing tests for other bindings can you add a separate group for testing `highlightOnFocus` and `autoFill`.

nick-woodward commented on src/ng-currency.directive.js:
```diff
@@ -146,13 +154,21 @@ export default function ngCurrency($filter, $locale, $timeout, ngCurrencySetting
       element.bind('focus', () => {
         if (active) {
+          if (autoFill && [undefined, null, ''].indexOf(controller.$$rawModelValue) !== -1) {
+            controller.$$rawModelValue = '0';
```
This makes it so once the user focuses the field the value is populated with zero (if its empty and autoFill is enabled).

Instead we should prepopulate it with zero.

cecilia-sanare reviewed (changes requested):
Thanks for your contribution @turt2live!

Overall your changes look good, just a couple of issues I noticed.

turt2live committed:
Add test for AutoFill

turt2live committed:
Make autofill populate on init instead of on focus

turt2live committed:
Add (failing) test for HighlightOnFocus

turt2live committed:
Update tests to hanle autoFill better.

HighlightOnFocus currently can't be handled due to a bug in phantomjs

nick-woodward commented on src/ng-currency.directive.js:
```diff
@@ -119,6 +127,10 @@ export default function ngCurrency($filter, $locale, $timeout, ngCurrencySetting
         controller.$validate();
         if (active) {
+          if (autoFill && [undefined, null, ''].indexOf(controller.$$rawModelValue) !== -1) {
+            controller.$$rawModelValue = '0';
```
One thing I noticed with this solution is that the value never gets commited to the `modelValue`.

Instead can you move this logic to the top of the `keepInRange` function. That way the if statement below gets executed.

```js
function keepInRange(value) {
  if (autoFill && [undefined, null, ''].indexOf(value) !== -1) {
    value = 0;
  }
  // ...
}
```

nick-woodward commented on test/ng-currency/ng-currency.directive.spec.js:
```diff
@@ -753,6 +755,51 @@ describe('ngCurrency directive tests', () => {
       });
     });
 
+    describe('AutoFill', () => {
```
Please validate the `scope.value` as well.

cecilia-sanare reviewed (changes requested):
I really like how you utilized `revalidate` to handle the `auto-fill`.

Once you resolve the final issue this PR should be good to go! :+1:

turt2live commented:
I'm also looking to add tri-state support to autoFill, as the original behaviour (on focus) is what we were using before in our own fork. Thoughts on having `autoFill` be `true`, `false`, or `focus`?

nick-woodward commented:
Sounds good to me!

turt2live committed:
wip on tri-state for autoFill (broken tests)

turt2live committed:
Validate scope.value in tests

turt2live committed:
More broken test for autoFill (min/max override checks)

I'm suspecting something is wrong with the actual test. More investigation required.

turt2live committed:
Fix all tests except for scope.value in focus autofill

turt2live committed:
Scope value not committed yet, so it shouldn't be treated as committed

turt2live committed:
Validate blur after focus to ensure the field returns to normal (failing)

turt2live committed:
Commit to view value if we autofilled it.

This is to ensure that the value actually gets saved when we blur.

turt2live commented:
@nick-woodward apologies for the long delay in getting updates through. This should now be complete.

Here's the rundown on how it's supposed to work (backed up by the tests):
* `autoFill` can be `true`, `false`, or `focus`
  * `true` - Pre-populate the field with `0` (bound by the min/max) on init
  * `false` - Default behaviour, do nothing different.
  * `focus` - Only pre-populate the field with `0` (bound by the min/max) on focus. Value intentionally retained on blur
* `highlightOnFocus` - Select the entire value on focus. Ideal for "quick edit" styled forms.

turt2live commented:
Unfortunately my spare time is limited and the project that was using this library doesn't need this patch any more. Going to have to close this :(
